### PR TITLE
[FSDE-24710] E2E tests for dropdown wc

### PR DESCRIPTION
**Pull request by soujanreddy**

### CODE REVIEW

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        cy.get(position.selector).click();
+        cy.get(position.selector).find(DROPDOWN_MENU)
+          .scrollIntoView({ offset: { top: -150, left: 0 } })
+          .should('have.attr', 'style')
+          .should('contain', `transform: ${position.transform}`);
+      })
+    });
+  });
+
+  describe.only('Methods', () => {
```
Comment (janibolkvadze) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:169]: .only

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+          'dropdownTrigger'
+        );
+        cy.get(DROPDOWN_DATA_CY.METHOD.TOGGLE.DROPDOWN_TOGGLE).find(DROPDOWN_MENU).as(
+          'dropdownMenu'
+        );
+      });
+      it('Should toggle the dropdown', () => {
+        cy.get(`@dropdownMenu`)
+          .should('not.be.visible');
+        cy.get(`@dropdownTrigger`).click();
```
Comment (janibolkvadze) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:182]: How does this check if the methods work? Seems like the implementation checks visibility toggling when clicking the standard trigger. The method is not called externally.

**Review by janibolkvadze — CHANGES_REQUESTED** [review]

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        cy.get(`@dropdownMenu`)
+          .should('not.be.visible');
+        cy.get(`@dropdownTrigger`).click();
+        cy.get(`@dropdownMenu`)
+          .should('not.be.visible');
+      });
+    });
+  });
+});
```
Comment (janibolkvadze) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:233]: + Event-related cases.

Comment (soujanreddy) [reply cypress/integration/chi-custom-elements/dropdown.spec.js:321]: Added

File: cypress/integration/chi-custom-elements/dropdown.spec.js (hunk 2 above)
Comment (soujanreddy) [reply cypress/integration/chi-custom-elements/dropdown.spec.js]: Updated

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+    it.only('Should open the dropdown', () => {
+      cy.get(`@dropdownMenu`)
+        .should('not.be.visible');
+      cy.get(`@dropdownTrigger`).click();
+      cy.get(`@dropdownMenu`)
+        .should('be.visible');
+    });
+
+    it.only('Should close the dropdown', () => {
```
[the remark this answers is not in the document]
Comment (soujanreddy) [reply cypress/integration/chi-custom-elements/dropdown.spec.js]: Nope. I have removed .only.

**Comment by soujanreddy** [comment]
@josecarlosgonzalezv Addressed updates as we discussed.

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+      cy.get('@dropdownTrigger').click().wait(500)
+      .then($els => {
+        const win = $els[0].ownerDocument.defaultView;
+        const after = win.getComputedStyle($els[0], ':after');
+        const tr = after.transform;
+        const values = tr.split('(')[1].split(')')[0].split(',');
+        const a = values[0];
+        const b = values[1];
+        const angle = Math.round(Math.atan2(b, a) * (180/Math.PI));
+        hasClassAssertion('@dropdownTrigger', ACTIVE_CLASS);
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:155]:
```suggestion
        const angle = Math.round(Math.atan2(b, a) * (180/Math.PI));
        
        hasClassAssertion('@dropdownTrigger', ACTIVE_CLASS);
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        );
+      });
+
+      it('Should toggle the dropdown menu when toggle method is called', () => {
+        cy.get(`@dropdownMenu`)
+          .should(('not.be.visible'));
+        cy.get(`@dropdown`)
+          .then(function(dropdown){
+          dropdown[0].toggle();
+          return new Promise(resolve => resolve(dropdown));
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:197]:
```suggestion
          .then(function(dropdown) {
            dropdown[0].toggle();
            return new Promise(resolve => resolve(dropdown));
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+          })
+          .find(DROPDOWN_MENU)
+          .should('be.visible');
+        cy.get(`@dropdown`)
+          .then(function(dropdown){
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:202]:
```suggestion
          .then(function(dropdown) {
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        );
+        cy.get(DROPDOWN_DATA_CY.METHOD.SHOW.DROPDOWN_SHOW).as(
+          'dropdown'
+        );
+      });
+      it('Should show the dropdown menu when show method is called', () => {
+        cy.get(`@dropdownMenu`)
+          .should(('not.be.visible'));
+        cy.get(`@dropdown`)
+          .then(function(dropdown){
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:228]:
```suggestion
          .then(function(dropdown) {
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+          'dropdownMenu'
+        );
+        cy.get(DROPDOWN_DATA_CY.EVENT.SHOW).as(
+          'dropdown'
+        );
+      });
+
+      it('Should trigger the appropriate hide event when the dropdown menu is opened', () => {
+        const spy = cy.spy();
+        cy.get('@dropdown').then(el => {
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:283]:
```suggestion
        const spy = cy.spy();
        
        cy.get('@dropdown').then(el => {
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+          'dropdownMenu'
+        );
+        cy.get(DROPDOWN_DATA_CY.EVENT.HIDE).as(
+          'dropdown'
+        );
+      });
+
+      it('Should trigger the appropriate hide event when the dropdown menu is hidden', () => {
+        const spy = cy.spy();
+        cy.get('@dropdown').then(el => {
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:315]:
```suggestion
        const spy = cy.spy();
        
        cy.get('@dropdown').then(el => {
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+      DROPDOWN_HIDE: '[data-cy="method-dropdown-hide"]',
+      TRIGGER: '#test-hide-trigger'
+    }
+  },
+  EVENT: {
+    SHOW: '[data-cy="event-dropdown-show"]',
+    HIDE: '[data-cy="event-dropdown-hide"]',
+  }
+};
+
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:81]:
```suggestion
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+      );
+      cy.get(DROPDOWN_DATA_CY.BASE).find(DROPDOWN_MENU).as(
+        'dropdownMenu'
+      );
+    });
+
+    it('Should open the dropdown', () => {
+      cy.get(`@dropdownMenu`)
+        .should('not.be.visible');
+      cy.get(`@dropdownTrigger`).click();
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:115]:
```suggestion
      cy.get(`@dropdownTrigger`)
        .click();
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+      cy.get(`@dropdownMenu`)
+        .should('be.visible');
+    });
+
+    it('Should close the dropdown', () => {
+      cy.get(`@dropdownMenu`)
+        .should('be.visible');
+      cy.get(`@dropdownTrigger`).click();
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:123]:
```suggestion
      cy.get(`@dropdownTrigger`)
        .click();
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+        .find(DROPDOWN_MENU).as(
+        'dropdownMenu'
+      );
+    });
+
+    it('Should rotate trigger chevron when toggling Dropdown visibility', () => {
+      cy.get(`@dropdownTrigger`)
+        .find(`.${ACTIVE_CLASS}`)
+        .should('not.exist');
+
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:145]:
```suggestion
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+      cy.get('@dropdownTrigger').click().wait(500)
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:146]:
```suggestion
      cy.get('@dropdownTrigger')
        .click()
        .wait(500)
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+          .should('be.visible');
+        expect(angle).to.equal(180);
+      });
+    });
+  });
+
+  describe('Position', () => {
+    DROPDOWN_DATA_CY.POSITION.forEach((position) => {
+      it(`Should be placed to the position ${position.placement}`, () => {
+        cy.get(position.selector).click();
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:166]:
```suggestion
        cy.get(position.selector)
          .click();
```

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 17]
```diff
@@ -0,0 +1,331 @@
...
+        cy.get(DROPDOWN_DATA_CY.METHOD.HIDE.DROPDOWN_HIDE).as(
+          'dropdown'
+        );
+      });
+
+      it('Should always hide the dropdown menu once it has been opened and closed', () => {
+        cy.get(`@dropdownMenu`)
+          .should('be.visible');
+        cy.get(`@dropdown`)
+          .then(function(dropdown){
```
Comment (josecarlosgonzalezv) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:255]:
```suggestion
          .then(function(dropdown) {
```

**Review by josecarlosgonzalezv — CHANGES_REQUESTED** [review]

**Review by josecarlosgonzalezv — APPROVED** [review]

File: cypress/integration/chi-custom-elements/dropdown.spec.js [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+const DROPDOWN_TRIGGER = '.chi-button.chi-dropdown__trigger';
+const DROPDOWN_MENU = '.chi-dropdown__menu';
+const ACTIVE_CLASS = '-active';
+const DONE_CLASS = '-done';
+const PENDING_CLASS = '-pending';
+const NUMBER_CLASS = 'chi-epanel__number';
+const BORDER_CLASS = '-bordered';
+const SLOT_CLASS_ACTIVE = 'chi-epanel__content-active';
+const SLOT_CLASS_DONE = 'done-slot';
+const SLOT_CLASS_FOOTER = 'chi-epanel__footer';
```
Comment (jllr) [inline cypress/integration/chi-custom-elements/dropdown.spec.js:90]:
```suggestion
```

**Review by jllr — CHANGES_REQUESTED** [review]

File: test/chi/custom-elements/dropdown.pug [hunk 19]
```diff
@@ -4,22 +4,44 @@ title: Dropdown
...
+.-mt--5(style='width: 500px; height: 500px;')
+  span.-text--h2=`Animate`
+  .-w--100.-h--100.-d--flex.-align-items--center.-justify-content--center
+    chi-dropdown(data-cy="animate-dropdown" active='false' button=`Animate` animate-chevron='true')
+      a.chi-dropdown__menu-item(href='#' slot='menu') Menu item 1
+      a.chi-dropdown__menu-item(href='#' slot='menu') Menu item 2
+
 - var positions = ['top', 'right', 'bottom', 'left', 'top-start', 'top-end', 'right-start', 'right-end', 'bottom-start', 'bottom-end', 'left-start', 'left-end'];
 - var methods = ['toggle', 'show', 'hide']
+- var events = ['show', 'hide']
```
Comment (jllr) [inline test/chi/custom-elements/dropdown.pug:30]: @soujanreddy in order to be consistent with the rest of tests we have in production, `vars` should be declared at the top of the page. 
Examples:
Date picker: https://github.com/CenturyLink/Chi/blob/master/test/chi/custom-elements/date-picker.pug
Drawer: https://github.com/CenturyLink/Chi/blob/master/test/chi/custom-elements/drawer.pug

**Review by jllr — APPROVED** [review]

**Pull request merged**